File: app/src/main/java/com/bigmikehoncho/mvvmdatabinding/MvvmBinder.java
package com.bigmikehoncho.mvvmdatabinding;

import android.app.Activity;
import android.databinding.DataBindingUtil;
import android.databinding.ViewDataBinding;
import android.support.annotation.NonNull;
import android.view.LayoutInflater;

import com.bigmikehoncho.mvvmdatabinding.adapters.ViewModelBinder;
import com.bigmikehoncho.mvvmdatabinding.utils.BindingUtils;
import com.bigmikehoncho.mvvmdatabinding.utils.Preconditions;

/**
 * Created by mike on 3/3/17.
 */

public class MvvmBinder<Binding extends ViewDataBinding> {
    private Binding binding;

    public MvvmBinder() {
    }

    /**
     * Call after super.onCreate of activity
     *
     * @param activity
     */
    public Binding onCreate(Activity activity) {
        VMBinder binder = (VMBinder) activity;
        binding = DataBindingUtil.setContentView(activity, binder.getLayoutId());
        getDefaultBinder().bind(binding, binder.createViewModel());
        return binding;
    }

    public Binding onCreate(VMBinder binder, LayoutInflater inflater) {
        binding = DataBindingUtil.inflate(inflater, binder.getLayoutId(), null, false);
        getDefaultBinder().bind(binding, binder.createViewModel());
        return binding;
    }

    /**
     * call before super.onDestroy or super.onDestroyView
     */
    public void onDestroy() {
        getDefaultBinder().bind(binding, null);
        binding.executePendingBindings();
    }

    @NonNull
    private ViewModelBinder getDefaultBinder() {
        ViewModelBinder defaultBinder = BindingUtils.getDefaultBinder();
        Preconditions.checkNotNull(defaultBinder, "Default Binder");
        return defaultBinder;
    }
}

File: mvvmDatabinding/src/main/java/com/bigmikehoncho/mvvmdatabinding/MvvmBinder.kt
package com.bigmikehoncho.mvvmdatabinding

import android.arch.lifecycle.Lifecycle
import android.arch.lifecycle.LifecycleObserver
import android.arch.lifecycle.OnLifecycleEvent
import android.arch.lifecycle.ViewModelProviders
import android.databinding.DataBindingUtil
import android.databinding.ViewDataBinding
import android.support.v7.app.AppCompatActivity
import android.util.Log
import com.bigmikehoncho.mvvmdatabinding.utils.BindingUtils

class MvvmBinder<out Binding : ViewDataBinding> : LifecycleObserver {

    private lateinit var binding: Binding
    private val defaultBinder = BindingUtils.defaultBinder

    /**
     * Call after super.createBinding of activity
     *
     * @param activity as VMBinder
     */
    fun createBinding(activity: AppCompatActivity): Binding {
        activity as VMBinder
        binding = DataBindingUtil.setContentView(activity, activity.layoutId)
        defaultBinder.bind(binding, ViewModelProviders.of(activity).get(activity.provideViewModelClass()))
        return binding
    }

    fun createBinding(fragment: android.support.v4.app.Fragment): Binding {
        fragment as VMBinder
        binding = DataBindingUtil.inflate(fragment.layoutInflater, fragment.layoutId, null, false)
        defaultBinder.bind(binding, ViewModelProviders.of(fragment).get(fragment.provideViewModelClass()))
        return binding
    }

    /**
     * Call before super.onDestroy or super.onDestroyView
     */
    @OnLifecycleEvent(Lifecycle.Event.ON_DESTROY)
    private fun onDestroy() {
        Log.i("MvvmBinder", "onDestroy")
        defaultBinder.bind(binding, null)
        binding.executePendingBindings()
    }
}
